Add discussion of ConfigureAwait

BillWagner opened the pull request:
The ConfigureAwait method can be used in async streams.

Fixes #13874

BillWagner committed:
Add discussion of ConfigureAwait

This can be used in async streams.

BillWagner committed:
fix xref build error

tdykstra reviewed (approved):
LGTM.  One suggestion for you to consider.

pkulikov commented on docs/csharp/language-reference/keywords/foreach-in.md:
```diff
@@ -19,7 +19,7 @@ The `foreach` statement executes a statement or a block of statements for each e
 Beginning with C# 7.3, if the enumerator's `Current` property returns a [reference return value](ref.md#reference-return-values) (`ref T` where `T` is the type of the collection element), you can declare the iteration variable with the `ref` or `ref readonly` modifier.
 
-Beginning with C# 8.0, the `await` operator may be applied to the `foreach` statement when the collection type implements the <xref:System.Collections.Generic.IAsyncEnumerable%601> interface. Each iteration of the loop may be suspended while the next element is retrieved asynchronously. 
+Beginning with C# 8.0, the `await` operator may be applied to the `foreach` statement when the collection type implements the <xref:System.Collections.Generic.IAsyncEnumerable%601> interface. Each iteration of the loop may be suspended while the next element is retrieved asynchronously. The extension method <xref:System.Threading.Tasks.TaskAsyncEnumerableExtensions.ConfigureAwait%2A?displayProper ...
```
@BillWagner @tdykstra should we move the focus from `ConfigureAwait` to context capturing? Something like that:

> By default, stream elements are processed in the captured context. If you want to disable capturing of the context, use the <xref:System.Threading.Tasks.TaskAsyncEnumerableExtensions.ConfigureAwait%2A?displayProperty=nameWithType> extension method.

Two questions to the above suggestion:
- Is the described default behavior correct?
- Is it possible to make a xref link with the specified argument value, something like:
  `<xref:System.Threading.Tasks.TaskAsyncEnumerableExtensions.ConfigureAwait(false)?displayProperty=nameWithType>` 
  to make it render as `TaskAsyncEnumerableExtensions.ConfigureAwait(false)`?

BillWagner replied:
> Is it possible to make a xref link with the specified argument value, something like:
<xref:System.Threading.Tasks.TaskAsyncEnumerableExtensions.ConfigureAwait(false)?displayProperty=nameWithType>
to make it render as TaskAsyncEnumerableExtensions.ConfigureAwait(false)?

No, that's not supported. (Great idea though)

BillWagner committed:
Update per feedback

- Rework sentence per @pkulikov.
- Add link to a general discussion of `ConfigureAwait` and captured contexts per @tdykstra

BillWagner committed:
off by one folder....

pkulikov commented on docs/csharp/language-reference/keywords/foreach-in.md:
```diff
@@ -19,7 +19,7 @@ The `foreach` statement executes a statement or a block of statements for each e
 Beginning with C# 7.3, if the enumerator's `Current` property returns a [reference return value](ref.md#reference-return-values) (`ref T` where `T` is the type of the collection element), you can declare the iteration variable with the `ref` or `ref readonly` modifier.
 
-Beginning with C# 8.0, the `await` operator may be applied to the `foreach` statement when the collection type implements the <xref:System.Collections.Generic.IAsyncEnumerable%601> interface. Each iteration of the loop may be suspended while the next element is retrieved asynchronously. 
+Beginning with C# 8.0, the `await` operator may be applied to the `foreach` statement when the collection type implements the <xref:System.Collections.Generic.IAsyncEnumerable%601> interface. Each iteration of the loop may be suspended while the next element is retrieved asynchronously. By default, stream elements are processed in the captured context. If you want to disable capturing of the cont ...
```
```suggestion
Beginning with C# 8.0, the `await` operator may be applied to the `foreach` statement when the collection type implements the <xref:System.Collections.Generic.IAsyncEnumerable%601> interface. Each iteration of the loop may be suspended while the next element is retrieved asynchronously. By default, stream elements are processed in the captured context. If you want to disable capturing of the context, use the <xref:System.Threading.Tasks.TaskAsyncEnumerableExtensions.ConfigureAwait%2A?displayProperty=nameWithType> extension method. For more information about synchronization contexts and capturing the current context, see the article on [consuming the Task-based asynchronous pattern](../../../standard/asynchronous-programming-patterns/consuming-the-task-based-asynchronous-pattern.md).
```

pkulikov commented on docs/csharp/tutorials/generate-consume-asynchronous-stream.md:
```diff
@@ -126,6 +126,8 @@ Replace that code with the following `await foreach` loop:
 
 [!code-csharp[FinishedEnumerateAsyncStream](~/samples/snippets/csharp/tutorials/AsyncStreams/finished/IssuePRreport/IssuePRreport/Program.cs#EnumerateAsyncStream)]
 
+By default, stream elements are processed in the captured context. If you want to disable capturing of the context, use the xref:System.Threading.Tasks.TaskAsyncEnumerableExtensions.ConfigureAwait%2A?displayProperty=nameWithType extension method call. For more information on synchronization contexts and capturing the current context, see the article on [consuming the Task based asynchronous patte ...
```
```suggestion
By default, stream elements are processed in the captured context. If you want to disable capturing of the context, use the <xref:System.Threading.Tasks.TaskAsyncEnumerableExtensions.ConfigureAwait%2A?displayProperty=nameWithType> extension method. For more information about synchronization contexts and capturing the current context, see the article on [consuming the Task-based asynchronous pattern](../../standard/asynchronous-programming-patterns/consuming-the-task-based-asynchronous-pattern.md).
```

pkulikov commented on docs/csharp/whats-new/csharp-8.md:
````diff
@@ -385,7 +385,7 @@ await foreach (var number in GenerateSequence())
 ```
 
-You can try asynchronous streams yourself in our tutorial on [creating and consuming async streams](../tutorials/generate-consume-asynchronous-stream.md).
+You can try asynchronous streams yourself in our tutorial on [creating and consuming async streams](../tutorials/generate-consume-asynchronous-stream.md). By default, stream elements are processed in the captured context. If you want to disable capturing of the context, use the xref:System.Threading.Tasks.TaskAsyncEnumerableExtensions.ConfigureAwait%2A?displayProperty=nameWithType extension metho ...
````
```suggestion
You can try asynchronous streams yourself in our tutorial on [creating and consuming async streams](../tutorials/generate-consume-asynchronous-stream.md). By default, stream elements are processed in the captured context. If you want to disable capturing of the context, use the <xref:System.Threading.Tasks.TaskAsyncEnumerableExtensions.ConfigureAwait%2A?displayProperty=nameWithType> extension method. For more information about synchronization contexts and capturing the current context, see the article on [consuming the Task-based asynchronous pattern](../../standard/asynchronous-programming-patterns/consuming-the-task-based-asynchronous-pattern.md).
```

pkulikov reviewed (commented):
Fixed xref links and minor style updates.

BillWagner committed:
Apply suggestions from code review

Co-Authored-By: Petr Kulikov <petr.kulikov@gmail.com>
